lib/MySQL_HostGroups_Manager.cpp

```diff
@@ -3419,23 +3476,45 @@ void MySQL_HostGroups_Manager::replication_lag_action_inner(MyHGC *myhgc, char *
 	}
 }
 
-void MySQL_HostGroups_Manager::replication_lag_action(int _hid, char *address, unsigned int port, int current_replication_lag) {
-	GloAdmin->mysql_servers_wrlock();
+void MySQL_HostGroups_Manager::replication_lag_action(const std::list<std::tuple<int, std::string, unsigned int, int>>& mysql_servers) {
+
+	//this method does not use admin table, so this lock is not needed. 
+	//GloAdmin->mysql_servers_wrlock();
+
+	int hid = -1;
+	std::string address;
+	unsigned int port = 0;
+	int current_replication_lag = -1;
+
+	unsigned long long curtime1 = monotonic_time();
 	wrlock();
-	if (mysql_thread___monitor_replication_lag_group_by_host == false) {
-		// legacy check. 1 check per server per hostgroup
-		MyHGC *myhgc = MyHGC_find(_hid);
-		replication_lag_action_inner(myhgc,address,port,current_replication_lag);
-	} else {
-		// only 1 check per server, no matter the hostgroup
-		// all hostgroups must be searched
-		for (unsigned int i=0; i<MyHostGroups->len; i++) {
-			MyHGC *myhgc=(MyHGC *)MyHostGroups->index(i);
-			replication_lag_action_inner(myhgc,address,port,current_replication_lag);
+
+	for (const auto& server : mysql_servers) {
+
+		std::tie(hid, address, port, current_replication_lag) = server;
```

Comment:
Same comment as placed for PR #4127. Here https://github.com/sysown/proxysql/pull/4127#discussion_r1149374997.